coderd/notifications/dispatch/smtp.go

```diff
@@ -260,10 +253,214 @@ func (s *SMTPHandler) dispatch(subject, htmlBody, plainBody, to string) Delivery
+	c.CommandTimeout = time.Until(deadline)
+	c.SubmissionTimeout = time.Until(deadline)
+
+	return c, nil
+}
+
+func (s *SMTPHandler) tlsConfig() (*tls.Config, error) {
+	host, _, err := s.smarthost()
+	if err != nil {
+		return nil, err
+	}
+
+	srvName := s.cfg.TLS.ServerName.String()
+	if srvName == "" {
+		srvName = host
+	}
+
+	ca, err := s.loadCAFile()
+	if err != nil {
+		return nil, xerrors.Errorf("load CA: %w", err)
+	}
+
+	var certs []tls.Certificate
+	cert, err := s.loadCertificate()
+	if err != nil {
+		return nil, xerrors.Errorf("load cert: %w", err)
+	}
+
+	if cert != nil {
+		certs = append(certs, *cert)
+	}
+
+	return &tls.Config{
+		ServerName: srvName,
+		// nolint:gosec // Users may choose to enable this.
+		InsecureSkipVerify: s.cfg.TLS.InsecureSkipVerify.Value(),
+
+		RootCAs:      ca,
+		Certificates: certs,
+		ClientAuth:   tls.RequireAndVerifyClientCert,
+	}, nil
+}
+
+func (s *SMTPHandler) loadCAFile() (*x509.CertPool, error) {
+	if s.cfg.TLS.CAFile == "" {
+		// nolint:nilnil // A nil CertPool is a valid response.
+		return nil, nil
+	}
+
+	ca, err := os.ReadFile(s.cfg.TLS.CAFile.String())
+	if err != nil {
+		return nil, xerrors.Errorf("load CA file: %w", err)
+	}
+
+	pool := x509.NewCertPool()
+	if !pool.AppendCertsFromPEM(ca) {
+		return nil, xerrors.Errorf("build cert pool: %w", err)
+	}
+
+	return pool, nil
+}
+
+func (s *SMTPHandler) loadCertificate() (*tls.Certificate, error) {
+	if len(s.cfg.TLS.CertFile) == 0 && len(s.cfg.TLS.KeyFile) == 0 {
+		// nolint:nilnil // A nil certificate is a valid response.
+		return nil, nil
+	}
+
+	cert, err := os.ReadFile(s.cfg.TLS.CertFile.Value())
+	if err != nil {
+		return nil, xerrors.Errorf("load cert: %w", err)
+	}
+	key, err := os.ReadFile(s.cfg.TLS.KeyFile.String())
+	if err != nil {
+		return nil, xerrors.Errorf("load key: %w", err)
+	}
+
+	pair, err := tls.X509KeyPair(cert, key)
+	if err != nil {
+		return nil, xerrors.Errorf("invalid or unusable keypair: %w", err)
+	}
+
+	return &pair, nil
+}
+
+// auth returns a value which implements the smtp.Auth based on the available auth mechanisms.
+func (s *SMTPHandler) auth(ctx context.Context, mechs string) (sasl.Client, error) {
+	username := s.cfg.Auth.Username.String()
+
+	var errs error
+	list := strings.Split(mechs, " ")
+	for _, mech := range list {
+		switch mech {
+		case sasl.Plain:
+			password, err := s.password()
+			if err != nil {
+				errs = multierror.Append(errs, err)
+				continue
+			}
+			if password == "" {
+				errs = multierror.Append(errs, xerrors.New("cannot use PLAIN auth, password not defined (see CODER_NOTIFICATIONS_EMAIL_AUTH_PASSWORD)"))
+				continue
+			}
+
+			return sasl.NewPlainClient(s.cfg.Auth.Identity.String(), username, password), nil
+		case sasl.Login:
+			if slices.Contains(list, sasl.Plain) {
+				// Prefer PLAIN over LOGIN.
+				continue
+			}
+
+			// Warn that LOGIN is obsolete, but don't do it every time we dispatch a notification.
+			loginWarnOnce.Do(func() {
```

Comment:
nit: Should we move it to initialization of the dispatcher?

Reply:
Good idea 👍